app-media-image-capture.html

```diff
@@ -391,20 +411,91 @@
+            // abstraction dedicated to managing constraints (or perhaps
+            // conflicts between controllers).
+            // @see https://www.w3.org/TR/mediacapture-streams/#dfn-applyconstraints
+            this._setTrackConstraints(trackConstraints);
+          });
+        },
+
+        __updatePhotoSettings: function() {
+          this.debounce('updatePhotoSettings', function() {
+            var photoSettings = this.__generateConfigurationObject(
+                this.photoCapabilities, PHOTO_SETTING_NAMES);
+
+            // Red eye reduction is a special case. The capability is reported
+            // as one of three states, and only one (controllable) allows the
+            // user to configure it.
+            // @see https://w3c.github.io/mediacapture-image/#redeyereduction-section
+            if (this.redEyeReduction != null &&
+                this.photoCapabilities.redEyeReduction === 'controllable') {
+              photoSettings = photoSettings || {};
+              photoSettings.redEyeReduction = this.redEyeReduction;
             }
 
-            this.imageCapture.setOptions(this.photoSettings)
-                .then(function() {
-                  this.fire('image-capture-options-updated', this);
-                }.bind(this))
-                .catch(function(e) {
-                  this.fire('image-capture-options-error', e);
-                }.bind(this));
+            this._setPhotoSettings(photoSettings);
           });
+        },
+
+        /**
+         * The purpose of this method is to take a set of capabilities reported
+         * by the platform, and sift the properties of this element through
+         * the filter of those reported capabilities. Properties that are not
+         * supported are ignored. Properties that are supported but don't have
+         * values within a MediaSettingsRange are clamped to that range.
+         * Properties with values that do not occur within a reported vector of
+         * supported values are ignored.
+         */
+        __generateConfigurationObject: function(reportedCapabilities, allowedNames) {
+          var configurationObject = null;
+
+          for (var i = 0; i < allowedNames.length; ++i) {
+            var name = allowedNames[i];
+            var value = this[name];
+            var capability = reportedCapabilities[name];
+            var capabilityIsArray = Array.isArray(capability);
+
+            // Don't set photo options if a value is provided, and skip
+            // if the name is not present in the PhotoCapabilities object
+            // (setting such properties will probably result in an error).
+            if (value == null ||
+                capability == null ||
+                capability === false ||
+                (capabilityIsArray && capability.indexOf(value) === -1)) {
+              continue;
+            }
+
+            if (configurationObject == null) {
```

Comment:
I was worried that an empty configuration object might cause problems. But... maybe not? I'll try it.